authenticator_idtoken.go

```diff
@@ -0,0 +1,54 @@
+package main
+
+import (
+	"net/http"
+
+	oidc "github.com/coreos/go-oidc"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+type idTokenAuthenticator struct {
+	header		string		// header name whiere id token is stored
+	caBundle	[]byte
+	provider	*oidc.Provider
+	clientID	string		// need client id to verify the id token
+	userIDClaim	string		// retrieve the userid if the claim exists
+}
+
+func (s *idTokenAuthenticator) AuthenticateRequest(r *http.Request) (*authenticator.Response, bool, error) {
+	logger := loggerForRequest(r)
+
+	// get id-token from header
+	bearer := getBearerToken(r.Header.Get(s.header))
+	if len(bearer) == 0 {
+		return nil, false, nil
+	}
+
+	ctx := setTLSContext(r.Context(), s.caBundle)
+
+	// Verifying received ID token
+	verifier := s.provider.Verifier(&oidc.Config{ClientID: s.clientID})
+	token, err := verifier.Verify(ctx, bearer)
+	if err != nil {
+		logger.Errorf("id-token verification failed: %v", err)
+		return nil, false, nil
+	}
+
+	// no error check here, because of the Verify() above
+	var claims map[string]interface{}
+	token.Claims(&claims)
+
+	if claims[s.userIDClaim] == nil {
+		// No USERID_CLAIM, pass this authenticator
+		logger.Error("USERID_CLAIM doesn't exist in the id token")
+		return nil, false, nil
+	}
+
```

Comment:
Could you also check for groups?
You can re-use this code: https://github.com/arrikto/oidc-authservice/blob/5f66e653d35f7f6b95ad9bffa7fd6019cc40abe0/server.go#L242-L246